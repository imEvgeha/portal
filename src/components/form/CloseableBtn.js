import React, {Component} from 'react';
import t from 'prop-types';
import {Button} from 'reactstrap';

export default class CloseableBtn extends Component {

    static propTypes = {
        title: t.string,
        id: t.string,
        value: t.string,
        disabled: t.bool,
        highlighted: t.bool,
        style: t.object,

        onClick: t.func,
        onClose: t.func,
    };

    constructor(props) {
        super(props);
        this.state = {};
        this.close = this.close.bind(this);
    }

    closeStyle = {
        fontWeight: '800',
        fontSize: '1.4rem',
        position: 'absolute',
        padding: '0 10px 0 2px',
        right: '0px',
        top: '0px'
    };

    close(event) {
        event.preventDefault();
        event.stopPropagation();
        this.props.onClose();
    }

    render() {
        return (
            <Button outline={!this.props.highlighted} color="secondary"
                    id={this.props.id}
                    onClick={this.props.onClick}
                    title={this.props.title + this.props.value}
                    disabled={this.props.disabled}
                    style={{width: '100%', textAlign: 'left', position: 'relative', ...this.props.style}}>
                <div style={{width: '100%', overflow: 'hidden', paddingRight: '20px', textOverflow: 'ellipsis'}}>
                    <small><strong>{this.props.title}</strong>{this.props.value}</small>
                </div>
                <a href={'#'} style={this.closeStyle} onClick={this.close} id={this.props.id + '-close-btn'}><span aria-hidden="true">&times;</span></a>
            </Button>
        );
    }
}
