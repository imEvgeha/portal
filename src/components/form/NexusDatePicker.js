import React, { Component } from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment';
import t from 'prop-types';
import {validateDate} from '../../util/Validation';
import {store} from '../../index';

const FIRST_DELIMETER_POSITION = 2;
const SECOUND_DELIMETER_POSITION = 5;

class NexusDatePicker extends Component {

    static propTypes = {
        date: t.any,
        id: t.string,
        disabled: t.bool,

        onChange: t.func,
        handleKeyPress: t.func,
        onInvalid: t.func,
    };

    prevRawInput = '';
    prevDate = null;

    constructor(props) {
        super(props);
        this.state = {
            dirty: false,
            open: false
        };

        this.clear = this.clear.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleChangeRaw = this.handleChangeRaw.bind(this);
        this.handleOnBlur = this.handleOnBlur.bind(this);
        this.handleClickOutside = this.handleClickOutside.bind(this);
        this.setValid = this.setValid.bind(this);
        this.refDatePicker = React.createRef();
    }

    toggleOpen = () => {
        if (this.state.open && this.refDatePicker.current.state.open) {
            setTimeout(() => {
                this.refDatePicker.current.setOpen(false);
            }, 10);
        }
        this.setState({open: !this.state.open});
    };

    componentDidMount() {
        this.refDatePicker.current.input.onclick = this.toggleOpen;
    }

    componentDidUpdate() {
        if (this.props.date === '') {
            this.prevDate = '';
            this.handleChange(null);
        }
        if (this.props.date === null && this.prevDate !== null) {
            this.prevDate = this.props.date;
            this.clear();
        }
        this.prevDate = this.props.date;
    }

    clear() {
        this.refDatePicker.current.clear();
        this.setState({dirty: false});
        this.prevRawInput = '';
        this.setValid(true);
    }

    handleChange(date) {
        this.props.onChange(date);
        this.setState({dirty: true});
        this.setValid(true);
    }

    handleOnBlur(locale = 'en') {
        this.setState({dirty: true, open: false});
        if (this.refDatePicker.current.input.value && !validateDate(this.refDatePicker.current.input.value, locale)) {
            this.setValid(false);
        }
    }

    // Hack for fix https://github.com/Hacker0x01/react-datepicker/issues/730
    handleClickOutside() {
        this.refDatePicker.current.cancelFocusInput();
        this.refDatePicker.current.setOpen(false);
    }

    setValid(valid) {
        this.setState({invalidDate: !valid});
        if (this.props.onInvalid) {
            this.props.onInvalid(!valid);
        }
    }

    // Used by ref
    focus() {
        this.refDatePicker.current.input.focus();
    }

    handleChangeRaw(date) {
        let justDirty = false;
        if (date) {
            if ( (this.prevRawInput.length === FIRST_DELIMETER_POSITION - 1 && date.length === FIRST_DELIMETER_POSITION) ||
                    (this.prevRawInput.length === SECOUND_DELIMETER_POSITION - 1 && date.length === SECOUND_DELIMETER_POSITION)) {
                this.refDatePicker.current.input.value = this.refDatePicker.current.input.value + '/';
            }

            if ( (this.prevRawInput.length === FIRST_DELIMETER_POSITION +1 && date.length === FIRST_DELIMETER_POSITION) ||
                    (this.prevRawInput.length === SECOUND_DELIMETER_POSITION + 1 && date.length === SECOUND_DELIMETER_POSITION || date.length > 10)) {
                this.refDatePicker.current.input.value = this.refDatePicker.current.input.value.slice(0, -1);
                date = this.refDatePicker.current.input.value;
            }

            if (!this.state.dirty && date.length === 10 ) {
                this.setState({dirty: true});
                justDirty = true;
            }

            if (validateDate(date)) {
                this.handleChange(moment(date));
            } else if (this.state.dirty || justDirty) {
                this.setValid(false);
            }

        } else {
            this.handleChange(null);
            this.setState({dirty: false});
        }
        this.prevRawInput = this.refDatePicker.current.input.value;
    }

    render() {
        const {locale} = store.getState().locale;

        return (
            <DatePicker
                ref={this.refDatePicker}
                className={this.state.invalidDate ? 'text-danger' : ''}
                id={this.props.id}
                placeholderText={'Enter date'}
                selected={this.props.date ? moment(this.props.date) : null}
                showYearDropdown
                showMonthDropdown
                autoComplete={'off'}
                allowSameDay={true}
                onChange={this.handleChange}
                onChangeRaw={(event) => this.handleChangeRaw(event.target.value)}
                onBlur={() => this.handleOnBlur(locale)}
                todayButton={'Today'}
                disabled={this.props.disabled}
                customInput={<input onKeyPress={this.props.handleKeyPress} />}
                onClickOutside={this.handleClickOutside}
                locale={locale}
            />
        );
    }
}

export default NexusDatePicker;
