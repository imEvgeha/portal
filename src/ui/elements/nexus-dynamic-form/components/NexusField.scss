@import '../../../../styles/atlaskit/mixins';
@import '../../../../styles/atlaskit/colors';

.nexus-c-field {
    padding: 0.75rem 1.25rem;
    margin-bottom: 5px;

    > div {
        display: flex;
    }

    .nexus-c-field__value-section {
        width: 100%;
    }

    .nexus-c-field__label {
        width: 25%;
        color: map-get($neutrals, N800);
    }

    &.nexus-c-field--error {
        background-color: rgb(242, 222, 222);
        color: rgb(169, 68, 66);

        .nexus-c-field__label {
            color: rgb(169, 68, 66);
        }
    }

    .nexus-c-field__placeholder {
        color: map-get($neutrals, N100);
    }

    .nexus-c-date-time-window-picker {
        padding: 0;
    }
}
