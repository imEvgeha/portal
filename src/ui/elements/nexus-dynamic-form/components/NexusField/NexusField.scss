@import '../../../../../styles/atlaskit/mixins';
@import '../../../../../styles/atlaskit/colors';

.nexus-c-field {
    padding: 0.75rem 1.25rem;
    margin-bottom: 5px;

    &.nexus-c-field--error {
        background-color: rgb(242, 222, 222);
        color: rgb(169, 68, 66);

        .nexus-c-field__label {
            color: rgb(169, 68, 66);
        }
    }

    .nexus-c-field__placeholder {
        color: map-get($neutrals, N100);
    }
}
