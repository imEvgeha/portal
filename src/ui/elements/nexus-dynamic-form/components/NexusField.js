import React from 'react';
import PropTypes from 'prop-types';
import {Checkbox} from '@atlaskit/checkbox';
import {DatePicker, TimePicker, DateTimePicker} from '@atlaskit/datetime-picker';
import {Field as AKField, ErrorMessage, CheckboxField} from '@atlaskit/form';
import TextField from '@atlaskit/textfield';
import NexusTextArea from "../../nexus-textarea/NexusTextArea";
import './NexusField.scss';

const NexusField = ({type, isEdit, tooltip, ...props}) => {
    const renderFieldEditMode = fieldProps => {
        switch (type) {
            case 'string':
                return <TextField {...fieldProps} />;
            case 'textarea':
                return <NexusTextArea {...fieldProps} />;
            case 'number':
                return <TextField {...fieldProps} type="Number" />;
            case 'date':
                return <DatePicker {...fieldProps} />;
            case 'time':
                return <TimePicker {...fieldProps} />;
            case 'datetime':
                return <DateTimePicker {...fieldProps} />;
            case 'boolean':
                const {name, label} = fieldProps;
                return (
                    <CheckboxField name={name} label={label}>
                        {({fieldProps}) => <Checkbox {...fieldProps} />}
                    </CheckboxField>
                );
            default:
                return;
        }
    };

    const renderFieldViewMode = fieldProps => {
        switch (type) {
            case 'datetime':
                return <DateTimePicker {...fieldProps} />;
            case 'boolean':
                return <Checkbox isDisabled defaultChecked={fieldProps.value} />;
            default:
                return <div>{fieldProps.value}</div>;
        }
    };

    return (
        <div className="nexus-c-field">
            <AKField {...props}>
                {({fieldProps, error, valid}) => (
                    <>
                        <div className="nexus-c-field__label">
                            {`${fieldProps.name}${fieldProps.isRequired ? '*' : ''}: `}
                            {tooltip && (
                                <span title={tooltip} style={{color: 'grey'}}>
                                    <i className="far fa-question-circle" />
                                </span>
                            )}
                        </div>
                        <div className="nexus-c-field__value">
                            {isEdit ? renderFieldEditMode(fieldProps) : renderFieldViewMode(fieldProps)}
                        </div>
                        {error && <ErrorMessage>{error}</ErrorMessage>}
                    </>
                )}
            </AKField>
        </div>
    );
};

NexusField.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    isEdit: PropTypes.bool,
    tooltip: PropTypes.string,
};

NexusField.defaultProps = {
    isEdit: false,
    tooltip: null,
};

export default NexusField;
