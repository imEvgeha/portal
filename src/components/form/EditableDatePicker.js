import React, { Component } from 'react';
import { Button } from 'reactstrap';
import FontAwesome from 'react-fontawesome';
import moment from 'moment';
import t from 'prop-types';
import NexusDatePicker from './NexusDatePicker';
import {INVALID_DATE} from '../../constants/messages';

class EditableDatePicker extends Component {

    static propTypes = {
        validate: t.func,
        value: t.string,
        displayName: t.string,
        disabled: t.bool,
        onChange: t.func,
        priorityDisplay: t.any
    };

    constructor(props) {
        const vodDate = props.value;
        super(props);
        this.state = {
            date: vodDate ? moment(vodDate) : moment(),
            datePickerStatus: false,
            errorMessage: '',
            submitStatus: false
        };

        this.handleShowDatePicker = this.handleShowDatePicker.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleCancelDatePicker = this.handleCancelDatePicker.bind(this);
        this.handleInvalid = this.handleInvalid.bind(this);
        this.submit = this.submit.bind(this);
        this.cancel = this.cancel.bind(this);
    }

    handleShowDatePicker(e) {
        e.preventDefault();
        if (!this.props.disabled) {
            this.setState({
                datePickerStatus: true,
                date: moment(this.state.date).isValid() ? moment(this.state.date) : moment(),
            });
        }
    }
    handleCancelDatePicker(e) {
        e.preventDefault();
        this.cancel();
    }

    cancel() {
        this.setState({
            date: moment(this.props.value),
            datePickerStatus: false,
            errorMessage: ''
        });
    }

    handleChange(date) {
        this.setState({ date: date, submitStatus: false, errorMessage: '' });
    }

    handleInvalid(invalid) {
        if (invalid) {
            this.setState({ errorMessage: INVALID_DATE, submitStatus: true });
        }
    }

    submit(date) {
        const validationError = this.props.validate(date);
        if (validationError !== undefined) {
            this.setState({
                errorMessage: validationError
            });
        } else {
            this.props.onChange(date, this.cancel);
            this.setState({
                datePickerStatus: false
            });
        }
    }

    render() {
        const displayFunc = (value)=>{
            return (<span
                       onClick={this.handleShowDatePicker}
                       className={'displayDate' + (this.props.disabled ? ' disabled' : '')}>
                       {value}
                   </span>);
        };

        const unfocusedRender = ()=>{
            if(this.props.priorityDisplay) {
                return displayFunc(this.props.priorityDisplay);
            } else {
                if(this.props.value) {
                    return displayFunc(this.state.date ? moment(this.state.date).format('L') : '');
                } else {
                    return(
                        <span
                            style={{ color: '#808080', cursor: 'pointer' }}
                            onClick={this.handleShowDatePicker}>
                            {this.props.disabled ? '' : 'Enter ' + this.props.displayName}
                        </span>
                    );
                }
            }
        };

        return (
            <div>
                {
                    this.state.datePickerStatus ?
                        <div>
                            <div className="dPicker" style={{ marginBottom: '5px' }}>
                                <NexusDatePicker
                                    id="dashboard-avails-search-start-date-text"
                                    date={this.state.date}
                                    onChange={this.handleChange}
                                    onInvalid={this.handleInvalid}
                                />
                            </div>
                            <div style={{ float: 'left' }}>
                                <Button
                                    className="dPButton"
                                    disabled={this.state.submitStatus}
                                    onClick={() => this.submit(this.state.date ? this.state.date.toISOString() : null)}
                                    color="success"><FontAwesome name='check' />
                                </Button>
                                <Button
                                    className="dPButton"
                                    onClick={this.handleCancelDatePicker}
                                    color="danger"><FontAwesome name='times' />
                                </Button>
                            </div>
                            {
                                this.state.errorMessage &&
                                <p style={{ color: 'red', float: 'left', width: '100%' }}>
                                    {this.state.errorMessage}
                                </p>
                            }
                        </div>
                        :
                        unfocusedRender()
                }
            </div>
        );
    }
}

export default EditableDatePicker;
