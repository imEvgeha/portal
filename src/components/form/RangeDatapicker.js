import t from 'prop-types';
import React from 'react';
import NexusDatePicker from './NexusDatePicker';
import {INVALID_DATE} from '../../constants/messages';
import moment from 'moment';
import {momentToISO} from '../../util/Common';

export default class RangeDatapicker extends React.Component {
    static propTypes = {
        id: t.string,
        value: t.object,
        displayName: t.string,
        disabled: t.bool,
        onFromDateChange: t.func,
        onToDateChange: t.func,
        onInvalid: t.func,
        handleKeyPress: t.func,
        hideLabel: t.bool
    };

    constructor(props) {
        super(props);
        this.state = {
            invalidStartDate: '',
            invalidEndDate: '',
            prevFromDate: null,
            prevToDate: null,
            invalidRange: ''
        };
        this.handleChangeStartDate = this.handleChangeStartDate.bind(this);
        this.handleChangeEndDate = this.handleChangeEndDate.bind(this);
        this.handleInvalid = this.handleInvalid.bind(this);
        this.refFirstDatePicker = React.createRef();
    }

    componentDidUpdate() {
        if (this.props.value.from && this.props.value.to && moment(this.props.value.from) > moment(this.props.value.to)) {
            if (!this.state.invalidRange && !this.state.invalidStartDate && !this.state.invalidEndDate) {
                this.props.onInvalid(this.wrongDateRange(true));
            }
        } else {
            if (this.state.invalidRange) {
                this.props.onInvalid(this.wrongDateRange(false));
            }
        }
    }

    handleChangeStartDate(date) {
        this.props.onFromDateChange(date ? momentToISO(date) : null);
        const invalidRange = this.wrongDateRange(date && this.props.value.to && moment(this.props.value.to) < date);
        this.setState({invalidStartDate: ''});
        this.props.onInvalid(this.state.invalidEndDate || invalidRange);
    }

    handleChangeEndDate(date) {
        this.props.onToDateChange(date ? momentToISO(date) : null);
        const invalidRange = this.wrongDateRange(date && this.props.value.from && moment(this.props.value.from) > date);
        this.setState({invalidEndDate: ''});
        this.props.onInvalid(this.state.invalidStartDate || invalidRange);
    }

    wrongDateRange(wrong) {
        const invalid = wrong ? this.props.displayName + ' from should be before to date' : '';
        this.setState({invalidRange: invalid});
        return !!invalid;
    }

    handleInvalid(name, value) {
        if (value) {
            this.setState({['invalid' + name +'Date']: INVALID_DATE, invalidRange: ''});
            this.props.onInvalid(true);
        } else {
            this.setState({['invalid' + name +'Date']: ''});
        }
    }

    focus() {
        this.refFirstDatePicker.current.focus();
    }

    render() {
        return (
            <div style={{ maxWidth:'300px', minWidth:'300px', flex:'1 1 300px', margin:'0 10px'}}>
                { !this.props.hideLabel && <label htmlFor="dashboard-avails-search-start-date-text">{this.props.displayName}</label>}
                <div className={'row justify-content-around'}>
                    <div style={{width: '45%', paddingLeft: '8px'}}>
                        <NexusDatePicker
                            id={this.props.id + '-from'}
                            date={this.props.value.from !== undefined ? this.props.value.from : ''}
                            ref={this.refFirstDatePicker}
                            onChange={this.handleChangeStartDate}
                            onInvalid={(value) => {this.handleInvalid('Start', value);}}
                            disabled={this.props.disabled}
                            handleKeyPress={this.props.handleKeyPress}
                        />
                        {this.state.invalidStartDate && <small className="text-danger ml-2"
                                                               style={{position: 'absolute'}}>{this.state.invalidStartDate}</small>}
                        {this.state.invalidRange && <small className="text-danger ml-2"
                                                           style={{position: 'absolute'}}>{this.state.invalidRange}</small>}
                    </div>
                    <div>_</div>
                    <div style={{width: '45%', paddingRight: '8px'}}>
                        <NexusDatePicker
                            id={this.props.id + '-to'}
                            date={this.props.value.to !== undefined ? this.props.value.to : ''}
                            onChange={this.handleChangeEndDate}
                            onInvalid={(value) => {this.handleInvalid('End', value);}}
                            disabled={this.props.disabled}
                            handleKeyPress={this.props.handleKeyPress}
                        />
                        {this.state.invalidEndDate && <small className="text-danger ml-2"
                                                               style={{position: 'absolute'}}>{this.state.invalidEndDate}</small>}
                    </div>
                </div>
            </div>
        );
    }
}
