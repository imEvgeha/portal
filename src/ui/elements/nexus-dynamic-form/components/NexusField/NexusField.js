import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {Checkbox} from '@atlaskit/checkbox';
import {Field as AKField, CheckboxField} from '@atlaskit/form';
import Select from '@atlaskit/select';
import TextField from '@atlaskit/textfield';
import {get, cloneDeep} from 'lodash';
import {compose} from 'redux';
import ErrorBoundary from '../../../../../pages/fallback/ErrorBoundary';
import NexusTextArea from '../../../nexus-textarea/NexusTextArea';
import {VIEWS} from '../../constants';
import withOptionalCheckbox from '../../hoc/withOptionalCheckbox';
import {
    checkFieldDependencies,
    getFieldValue,
    getValidationFunction,
    formatOptions,
    renderLabel,
    renderError,
} from '../../utils';
import CastCrew from './components/CastCrew/CastCrew';
import DateTime from './components/DateTime/DateTime';
import './NexusField.scss';

const DateTimeWithOptional = compose(withOptionalCheckbox())(DateTime);

const NexusTextAreaWithOptional = compose(withOptionalCheckbox())(NexusTextArea);

const CheckboxWithOptional = compose(withOptionalCheckbox())(Checkbox);

const SelectWithOptional = compose(withOptionalCheckbox())(Select);

const TextFieldWithOptional = compose(withOptionalCheckbox())(TextField);

const NexusField = ({
    isHighlighted,
    selectValues,
    path,
    type,
    view,
    tooltip,
    formData,
    isReadOnly,
    isRequired,
    dependencies,
    validationError,
    validation,
    dateType,
    labels,
    optionsConfig,
    label,
    isOptional,
    setFieldValue,
    useCurrentDate,
    ...props
}) => {
    const [fetchedOptions, setFetchedOptions] = useState([]);

    useEffect(() => {
        if ((type === 'select' || type === 'multiselect') && optionsConfig.options === undefined) {
            if (Object.keys(selectValues).length) {
                let options = cloneDeep(selectValues[path]);
                options = formatOptions(options, optionsConfig);
                setFetchedOptions(options);
            }
        }
    }, [selectValues]);

    const checkDependencies = type => {
        return checkFieldDependencies(type, view, dependencies, formData);
    };

    const addedProps = {
        isOptional,
        setFieldValue,
        path,
        view,
    };

    const dateProps = {
        labels,
        type,
        dateType,
        isReadOnly,
        useCurrentDate,
        ...addedProps,
    };

    const renderFieldEditMode = fieldProps => {
        const {options} = optionsConfig;
        const selectFieldProps = {...fieldProps};
        const multiselectFieldProps = {...fieldProps};
        switch (type) {
            case 'string':
            case 'stringInArray':
                return <TextFieldWithOptional {...fieldProps} placeholder={`Enter ${label}`} {...addedProps} />;
            case 'textarea':
                return <NexusTextAreaWithOptional {...fieldProps} placeholder={`Enter ${label}`} {...addedProps} />;
            case 'number':
                return (
                    <TextFieldWithOptional
                        {...fieldProps}
                        type="Number"
                        placeholder={`Enter ${label}`}
                        {...addedProps}
                    />
                );
            case 'boolean':
                return (
                    <CheckboxField
                        isDisabled={isReadOnly || checkDependencies('readOnly')}
                        name={fieldProps.name}
                        label={fieldProps.label}
                        defaultIsChecked={fieldProps.value}
                    >
                        {({fieldProps}) => <CheckboxWithOptional {...addedProps} {...fieldProps} />}
                    </CheckboxField>
                );
            case 'select':
                if (get(fieldProps, 'value.value', undefined) === undefined) {
                    selectFieldProps.value = {
                        label: fieldProps.value,
                        value: fieldProps.value,
                    };
                }
                return (
                    <SelectWithOptional
                        {...selectFieldProps}
                        options={options !== undefined ? options : fetchedOptions}
                        defaultValue={fieldProps.value ? {value: fieldProps.value, label: fieldProps.value} : undefined}
                        {...addedProps}
                    />
                );
            case 'multiselect':
                if (
                    fieldProps.value &&
                    fieldProps.value.length &&
                    fieldProps.value[fieldProps.value.length - 1].value === undefined
                ) {
                    multiselectFieldProps.value = fieldProps.value.map(val => ({label: val, value: val}));
                }
                return (
                    <SelectWithOptional
                        {...multiselectFieldProps}
                        options={options !== undefined ? options : fetchedOptions}
                        isMulti
                        defaultValue={
                            fieldProps.value
                                ? fieldProps.value.map(val => {
                                      return {label: val, value: val};
                                  })
                                : undefined
                        }
                        {...addedProps}
                    />
                );
            case 'dateRange':
            case 'datetime':
                return <DateTimeWithOptional {...dateProps} {...fieldProps} />;
            case 'castCrew':
                return <CastCrew {...fieldProps} persons={fieldProps.value ? fieldProps.value : []} isEdit={true} />;
            default:
                return;
        }
    };

    const getValue = fieldProps => {
        if (Array.isArray(fieldProps.value)) {
            if (fieldProps.value.length) {
                return fieldProps.value.map(x => x && getFieldValue(x)).join(', ');
            }
            return <div className="nexus-c-field__placeholder">{`Enter ${label}...`}</div>;
        }
        return getFieldValue(fieldProps.value);
    };

    const renderFieldViewMode = fieldProps => {
        if (validationError) {
            return <div>{validationError}</div>;
        }
        switch (type) {
            case 'boolean':
                return <Checkbox isDisabled defaultChecked={fieldProps.value} />;
            case 'dateRange':
            case 'datetime':
                if (fieldProps.value) {
                    return <DateTime {...dateProps} {...fieldProps} isReadOnly />;
                }
                return <div className="nexus-c-field__placeholder">{`Enter ${label}...`}</div>;
            case 'castCrew':
                return <CastCrew persons={fieldProps.value ? fieldProps.value : []} isEdit={false} />;
            default:
                return fieldProps.value ? (
                    <div>{getValue(fieldProps)}</div>
                ) : (
                    <div className="nexus-c-field__placeholder">{`Enter ${label}...`}</div>
                );
        }
    };

    const required = !!(checkDependencies('required') || isRequired);
    return (
        <ErrorBoundary>
            <div
                className={`nexus-c-field${validationError ? ' nexus-c-field--error' : ''}${
                    isHighlighted ? ' nexus-c-field--highlighted' : ''
                }`}
            >
                <AKField
                    isDisabled={isReadOnly || checkDependencies('readOnly')}
                    isRequired={checkDependencies('required') || isRequired}
                    validate={value => getValidationFunction(value, validation, {type, isRequired: required})}
                    {...props}
                >
                    {({fieldProps, error}) => (
                        <>
                            {type !== 'castCrew' &&
                                renderLabel(label, !!(checkDependencies('required') || isRequired), tooltip)}
                            <div className="nexus-c-field__value-section">
                                <div className="nexus-c-field__value">
                                    {view === VIEWS.EDIT || view === VIEWS.CREATE
                                        ? renderFieldEditMode(fieldProps)
                                        : renderFieldViewMode(fieldProps)}
                                </div>
                                {renderError({...fieldProps}, error)}
                            </div>
                        </>
                    )}
                </AKField>
            </div>
        </ErrorBoundary>
    );
};

NexusField.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    view: PropTypes.string,
    tooltip: PropTypes.string,
    formData: PropTypes.object,
    dependencies: PropTypes.array,
    isReadOnly: PropTypes.bool,
    isRequired: PropTypes.bool,
    validationError: PropTypes.string,
    validation: PropTypes.array,
    optionsConfig: PropTypes.object,
    selectValues: PropTypes.object,
    path: PropTypes.any,
    dateType: PropTypes.string,
    labels: PropTypes.array,
    label: PropTypes.string,
    isOptional: PropTypes.bool,
    setFieldValue: PropTypes.func,
    // eslint-disable-next-line react/boolean-prop-naming
    useCurrentDate: PropTypes.bool,
    isHighlighted: PropTypes.bool,
};

NexusField.defaultProps = {
    view: VIEWS.VIEW,
    tooltip: null,
    formData: {},
    dependencies: [],
    isReadOnly: false,
    isRequired: false,
    validationError: null,
    validation: [],
    optionsConfig: {},
    selectValues: {},
    path: null,
    dateType: '',
    labels: [],
    label: '',
    isOptional: false,
    setFieldValue: null,
    useCurrentDate: false,
    isHighlighted: false,
};

export default NexusField;
