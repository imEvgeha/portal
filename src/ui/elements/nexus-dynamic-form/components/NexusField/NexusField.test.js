import React from 'react';
import {Field as AKField} from '@atlaskit/form';
import {shallow} from 'enzyme';
import NexusField from './NexusField';

describe('NexusField', () => {
    let wrapper = null;
    const props = {
        key: 'rightId',
        name: 'rightId',
        label: 'Right ID',
        type: 'string',
        isDisabled: true,
        defaultValue: 'rght_zrp8g',
    };

    beforeEach(() => {
        wrapper = shallow(<NexusField {...props} />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a AKField', () => {
        const textField = wrapper.find(AKField);
        expect(textField.length).toEqual(1);
    });
});
