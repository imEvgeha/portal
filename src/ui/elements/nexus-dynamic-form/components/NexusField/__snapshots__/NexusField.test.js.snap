// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusField should match snapshot 1`] = `
<ErrorBoundary>
  <div
    className="nexus-c-field "
  >
    <Field
      defaultValue="rght_zrp8g"
      isDisabled={true}
      isRequired={false}
      name="rightId"
      validate={[Function]}
    >
      <Component />
    </Field>
  </div>
</ErrorBoundary>
`;
